pennylane/decomposition/decomposition_graph.py

```diff
@@ -160,29 +168,75 @@ def _recursively_add_op_node(self, op_node: CompressedResourceOp) -> int:
 
         op_node_idx = self._graph.add_node(op_node)
         self._op_node_indices[op_node] = op_node_idx
+
         if op_node.op_type.__name__ in self._target_gate_set:
             self._target_gate_indices.add(op_node_idx)
             return op_node_idx
 
+        if op_node.op_type in (qml.ops.Controlled, qml.ops.ControlledOp):
+            # This branch only applies to general controlled operators
+            return self._add_controlled_decomp_node(op_node, op_node_idx)
+
         for decomposition in self._get_decompositions(op_node.op_type):
-            d_node_idx = self._recursively_add_decomposition_node(decomposition, op_node.params)
+            decomp_resource = decomposition.compute_resources(**op_node.params)
+            d_node_idx = self._recursively_add_decomposition_node(decomposition, decomp_resource)
             self._graph.add_edge(d_node_idx, op_node_idx, 0)
 
         return op_node_idx
 
-    def _recursively_add_decomposition_node(self, rule: DecompositionRule, params: dict) -> int:
+    def _add_special_decomp_rule_to_op(
+        self, rule: DecompositionRule, op_node: CompressedResourceOp, op_node_idx: int
+    ):
+        """Adds a special decomposition rule to the graph."""
+        decomp_resource = rule.compute_resources(**op_node.params)
+        d_node_idx = self._recursively_add_decomposition_node(rule, decomp_resource)
+        self._graph.add_edge(d_node_idx, op_node_idx, 0)
+
+    def _add_controlled_decomp_node(self, op_node: CompressedResourceOp, op_node_idx: int) -> int:
+        """Adds a controlled decomposition node to the graph."""
+
+        base_class = op_node.params["base_class"]
+        num_control_wires = op_node.params["num_control_wires"]
+
+        # Handle controlled global phase
+        if base_class is qml.GlobalPhase:
+            rule = controlled_global_phase_decomp
+            self._add_special_decomp_rule_to_op(rule, op_node, op_node_idx)
+            return op_node_idx
+
+        # Handle controlled-X gates
+        if base_class is qml.X:
+            rule = controlled_x_decomp
+            self._add_special_decomp_rule_to_op(rule, op_node, op_node_idx)
+            return op_node_idx
+
+        # Handle custom controlled ops
+        if (base_class, num_control_wires) in base_to_custom_ctrl_op():
+            custom_op_type = base_to_custom_ctrl_op()[(base_class, num_control_wires)]
+            rule = CustomControlledDecomposition(custom_op_type)
+            self._add_special_decomp_rule_to_op(rule, op_node, op_node_idx)
+            return op_node_idx
+
+        # General case
+        for base_decomposition in self._get_decompositions(base_class):
+            rule = ControlledBaseDecomposition(base_decomposition)
+            self._add_special_decomp_rule_to_op(rule, op_node, op_node_idx)
+
+        return op_node_idx
```

Comment:
Wait, am I missing something or can we pull the `_add_special_decomp_rule_to_op` and the return statement out of the if clauses? They are the same, no?

Reply:
No, there's a return statement in each branch. If one branch is satisfied, we execute that branch and return right away. I guess we could turn the `if`s into `elif`s, but then we would have to put the general case (the for loop) in an `else`, but then we would still need to return from the `else`. Basically I think the current structure is easier to read.

Reply:
Okay, if it is a conscious decision, fine by me :)